modal/io_streams.py

```diff
@@ -89,12 +98,25 @@ def __init__(
         self._client = client
         self._stream = None
         self._last_entry_id = None
-        self._buffer = ""
+        self._line_buffer = ""
         self._by_line = by_line
         # Whether the reader received an EOF. Once EOF is True, it returns
         # an empty string for any subsequent reads (including async for)
         self.eof = False
 
+        # We only support piping sandbox logs because they're meant to be durable logs stored
+        # on the user's application.
+        if object_type == "sandbox" and stream_type != StreamType.PIPE:
+            raise ValueError("Sandbox streams must be piped.")
+        self._stream_type = stream_type
+
+        if self._object_type == "container_process":
+            # Container process streams need to be consumed as they are produced,
+            # otherwise the process will block. Use a buffer to store the stream
+            # until the client consumes it.
+            self._container_process_buffer = []
+            asyncio.create_task(self._consume_container_process_stream())
```

Comment:
You need to keep a reference to this task, otherwise it could get preemptively garbage collected and just stop. Also, it should preferably be created in an async context manager that can be entered/exited by some "owner" of this stream, so we don't "leak" this task in case of cancellations/early returns (garbage collection of async tasks can't be trusted to execute `finally`-statements etc. of those tasks since the gc happens outside of the event loop)

Reply:
I just realized the context manager thing might be tricky to pull of w/ how we currently structure sandbox calls to implicitly just start streaming logs in the background when created.

That part could be left as a future improvement where we make ContainerProcess into an (optional) context manager, which could allow graceful termination of the underlying streams (and even the process itself) when exited.